src/toil_lib/programs.py

```diff
@@ -74,10 +84,17 @@ def docker_call(tool,
                     outfile = download_url(url, work_dir=work_dir, name=filename)
                 assert os.path.exists(file_path)
         return
-    
+
+    container_name = _get_container_name(job)
     base_docker_call = ['docker', 'run',
                         '--log-driver=none',
                         '-v', '{}:/data'.format(os.path.abspath(work_dir))]
+
+    # Defer the permission fixing function.  We call this explicitly later on in this function, but
+    # we defer it as well to handle unexpected job failure.
+    job.defer(_fix_permissions, base_docker_call, tool, work_dir)
+
+    base_docker_call.extend(['--name', container_name])
     if rm:
         base_docker_call.append('--rm')
```

Comment:
So `--rm` is apparently unreliable because it is implemented in the client, not the daemon. 

https://github.com/docker/docker/pull/20848

We can't be certain that a container started with `--rm` will be removed when you `docker kill` it. I think we should add an explicit `docker rm` to `_docker_kill` if `rm=True` was passed to this function.